### PR TITLE
add various coments to explain how the TAIT code works

### PR DESCRIPTION
**Pull request by nikomatsakis**
r? @oli-obk

### CODE REVIEW

**Review by spastorino — APPROVED** [review]

File: compiler/rustc_mir/src/borrow_check/type_check/mod.rs [hunk 1]
```diff
@@ -1206,6 +1206,35 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
...
+    /// Here, the function signature would be something like
+    /// `fn(T) -> Box<impl Debug>`. The MIR return slot would have
+    /// the type with the opaque type revealed, so `Box<(T, u32)>`.
+    ///
+    /// In terms of our function parameters:
+    ///
+    /// * `anon_ty` would be `Box<Foo<T>>` where `Foo` is an opaque type
+    ///   scoped to this function (note that it is parameterized by the
+    ///   generics of `foo`).
+    /// * `revealed_ty` would be `Box<(Foo<T>, u32)>`
```
Comment (oli-obk) [inline compiler/rustc_mir/src/borrow_check/type_check/mod.rs:1236]:
```suggestion
    /// * `revealed_ty` would be `Box<(T, u32)>`
```

File: compiler/rustc_mir/src/borrow_check/type_check/mod.rs [hunk 2]
```diff
@@ -1206,6 +1206,35 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
...
+    /// }
+    /// ```
+    ///
+    /// Here, the function signature would be something like
+    /// `fn(T) -> Box<impl Debug>`. The MIR return slot would have
+    /// the type with the opaque type revealed, so `Box<(T, u32)>`.
+    ///
+    /// In terms of our function parameters:
+    ///
+    /// * `anon_ty` would be `Box<Foo<T>>` where `Foo` is an opaque type
```
Comment (oli-obk) [inline compiler/rustc_mir/src/borrow_check/type_check/mod.rs:1233]: Isn't the anon_ty just `Foo<T>`?

File: compiler/rustc_mir/src/borrow_check/type_check/mod.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/borrow_check/type_check/mod.rs:1236]: and then this would actually just be `(T, u32)`

**Review by oli-obk — CHANGES_REQUESTED** [review]
Something in the docs doesn't add up 100%, r=me with that fixed

File: compiler/rustc_mir/src/borrow_check/type_check/mod.rs (hunk 2 above)
Comment (nikomatsakis) [reply compiler/rustc_mir/src/borrow_check/type_check/mod.rs:1233]: No, it's not. It's the return type, which *contains* opaque types (but also maybe other stuff). That's explicitly why I included the `Box` in the example. Let me clarify that.

File: compiler/rustc_mir/src/borrow_check/type_check/mod.rs (hunk 1 above)
Comment (nikomatsakis) [reply compiler/rustc_mir/src/borrow_check/type_check/mod.rs:1236]: correct

**Pull request merged**